In [1]:
import pandas as pd

In [2]:
df=pd.read_csv(r"archive/sms-call-internet-mi-2013-11-01.csv")
df

,datetime,CellID,countrycode,smsin,smsout,callin,callout,internet
0,2013-11-01 00:00:00,1,0,0.3521,NaN,NaN,0.0273,NaN
1,2013-11-01 00:00:00,1,33,NaN,NaN,NaN,NaN,0.0261
2,2013-11-01 00:00:00,1,39,1.7322,1.1047,0.5919,0.4020,57.7729
3,2013-11-01 00:00:00,2,0,0.3581,NaN,NaN,0.0273,NaN
4,2013-11-01 00:00:00,2,33,NaN,NaN,NaN,NaN,0.0274
...,...,...,...,...,...,...,...,...
1891923,2013-11-01 23:00:00,10000,20,0.0860,NaN,NaN,NaN,NaN
1891924,2013-11-01 23:00:00,10000,33,0.0542,NaN,NaN,NaN,NaN
1891925,2013-11-01 23:00:00,10000,39,3.3318,3.6637,0.8509,1.0462,99.3640
1891926,2013-11-01 23:00:00,10000,49,NaN,NaN,NaN,0.0048,NaN


In [3]:
print(df.shape)

(1891928, 8)


In [4]:
print(df.columns.tolist())

['datetime', 'CellID', 'countrycode', 'smsin', 'smsout', 'callin', 'callout', 'internet']


In [5]:
df.dtypes

datetime           str
CellID           int64
countrycode      int64
smsin          float64
smsout         float64
callin         float64
callout        float64
internet       float64
dtype: object

In [6]:
df.head()


,datetime,CellID,countrycode,smsin,smsout,callin,callout,internet
0,2013-11-01 00:00:00,1,0,0.3521,NaN,NaN,0.0273,NaN
1,2013-11-01 00:00:00,1,33,NaN,NaN,NaN,NaN,0.0261
2,2013-11-01 00:00:00,1,39,1.7322,1.1047,0.5919,0.4020,57.7729
3,2013-11-01 00:00:00,2,0,0.3581,NaN,NaN,0.0273,NaN
4,2013-11-01 00:00:00,2,33,NaN,NaN,NaN,NaN,0.0274


In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1891928 entries, 0 to 1891927
Data columns (total 8 columns):
 #   Column       Dtype  
---  ------       -----  
 0   datetime     str    
 1   CellID       int64  
 2   countrycode  int64  
 3   smsin        float64
 4   smsout       float64
 5   callin       float64
 6   callout      float64
 7   internet     float64
dtypes: float64(5), int64(2), str(1)
memory usage: 115.5 MB


In [8]:
df=df.rename(columns={
    "datetime":"timestamp",
"CellID":"grid_id",
"countrycode":"country_code",
"smsin":"sms_in",
"smsout":"sms_out",
"callin":"call_in",
"callout":"call_out", 
"internet":"internet_activity" 
})
print(df.columns.tolist())

['timestamp', 'grid_id', 'country_code', 'sms_in', 'sms_out', 'call_in', 'call_out', 'internet_activity']


In [9]:
df["timestamp"] = pd.to_datetime(df["timestamp"])

In [10]:
df["timestamp"].unique()

<DatetimeArray>
['2013-11-01 00:00:00', '2013-11-01 01:00:00', '2013-11-01 02:00:00',
 '2013-11-01 03:00:00', '2013-11-01 04:00:00', '2013-11-01 05:00:00',
 '2013-11-01 06:00:00', '2013-11-01 07:00:00', '2013-11-01 08:00:00',
 '2013-11-01 09:00:00', '2013-11-01 10:00:00', '2013-11-01 11:00:00',
 '2013-11-01 12:00:00', '2013-11-01 13:00:00', '2013-11-01 14:00:00',
 '2013-11-01 15:00:00', '2013-11-01 16:00:00', '2013-11-01 17:00:00',
 '2013-11-01 18:00:00', '2013-11-01 19:00:00', '2013-11-01 20:00:00',
 '2013-11-01 21:00:00', '2013-11-01 22:00:00', '2013-11-01 23:00:00']
Length: 24, dtype: datetime64[us]

In [11]:
df["timestamp"].nunique()

24

In [12]:
timestamps = df["timestamp"].dropna().sort_values().unique()

print(len(timestamps))
print(timestamps)

24
<DatetimeArray>
['2013-11-01 00:00:00', '2013-11-01 01:00:00', '2013-11-01 02:00:00',
 '2013-11-01 03:00:00', '2013-11-01 04:00:00', '2013-11-01 05:00:00',
 '2013-11-01 06:00:00', '2013-11-01 07:00:00', '2013-11-01 08:00:00',
 '2013-11-01 09:00:00', '2013-11-01 10:00:00', '2013-11-01 11:00:00',
 '2013-11-01 12:00:00', '2013-11-01 13:00:00', '2013-11-01 14:00:00',
 '2013-11-01 15:00:00', '2013-11-01 16:00:00', '2013-11-01 17:00:00',
 '2013-11-01 18:00:00', '2013-11-01 19:00:00', '2013-11-01 20:00:00',
 '2013-11-01 21:00:00', '2013-11-01 22:00:00', '2013-11-01 23:00:00']
Length: 24, dtype: datetime64[us]


In [13]:
time_diffs = pd.Series(timestamps).diff()

print(time_diffs.value_counts())
print((time_diffs.dropna() == pd.Timedelta(hours=1)).all())

0 days 01:00:00    23
Name: count, dtype: int64
True


In [14]:
assert df["timestamp"].nunique() == 24

In [15]:
df["date"] = df["timestamp"].dt.date
df["hour"] = df["timestamp"].dt.hour
df["day_of_week"] = df["timestamp"].dt.day_name()

In [16]:
print(df[["timestamp", "date", "hour", "day_of_week"]].head())

   timestamp        date  hour day_of_week
0 2013-11-01  2013-11-01     0      Friday
1 2013-11-01  2013-11-01     0      Friday
2 2013-11-01  2013-11-01     0      Friday
3 2013-11-01  2013-11-01     0      Friday
4 2013-11-01  2013-11-01     0      Friday


In [17]:
print("Missing grid_id:", df["grid_id"].isna().sum())
print("Missing timestamp:", df["timestamp"].isna().sum())

Missing grid_id: 0
Missing timestamp: 0


In [18]:
print(df.isna().sum())

timestamp                  0
grid_id                    0
country_code               0
sms_in               1086153
sms_out              1422446
call_in              1407781
call_out             1037413
internet_activity    1087074
date                       0
hour                       0
day_of_week                0
dtype: int64


In [19]:
activity_cols = [
    "sms_in",
    "sms_out",
    "call_in",
    "call_out",
    "internet_activity"
]
print(df[activity_cols].isna().sum())

sms_in               1086153
sms_out              1422446
call_in              1407781
call_out             1037413
internet_activity    1087074
dtype: int64


In [20]:
duplicate_count = df.duplicated().sum()

print("Exact duplicates:", duplicate_count)

Exact duplicates: 0


In [39]:
print("Negative Values:")
for col in activity_cols:
    print(col, (df[col] < 0).sum())

Negative Values:
sms_in 0
sms_out 0
call_in 0
call_out 0
internet_activity 0


In [22]:
print("Minimum grid_id:", df["grid_id"].min())
print("Maximum grid_id:", df["grid_id"].max())
print("Unique grids:", df["grid_id"].nunique())

Minimum grid_id: 1
Maximum grid_id: 10000
Unique grids: 10000


In [23]:
invalid_grids = df[
    (df["grid_id"] < 1) |
    (df["grid_id"] > 10000)
]

print(invalid_grids)

Empty DataFrame
Columns: [timestamp, grid_id, country_code, sms_in, sms_out, call_in, call_out, internet_activity, date, hour, day_of_week]
Index: []


In [24]:
print(df["country_code"].value_counts())
print(df["country_code"].nunique())

country_code
39      240000
0       231307
46      109848
33      106603
49       95933
         ...  
1938         4
247          3
240          2
596          2
1709         2
Name: count, Length: 246, dtype: int64
246


In [25]:
grid_hour_counts = (
    df.groupby(["timestamp", "grid_id"])
      .size()
      .reset_index(name="row_count")
)

print(grid_hour_counts.head())

   timestamp  grid_id  row_count
0 2013-11-01        1          3
1 2013-11-01        2          3
2 2013-11-01        3          3
3 2013-11-01        4          3
4 2013-11-01        5          3


In [26]:
example = df[
    (df["timestamp"] == "2013-11-01 00:00:00") &
    (df["grid_id"] == 1)
]

print(example[["timestamp", "grid_id", "country_code"]])

   timestamp  grid_id  country_code
0 2013-11-01        1             0
1 2013-11-01        1            33
2 2013-11-01        1            39


In [27]:
examples = grid_hour_counts[
    grid_hour_counts["row_count"] > 1
]

print(examples.head())

example = examples.iloc[0]

print(example)

   timestamp  grid_id  row_count
0 2013-11-01        1          3
1 2013-11-01        2          3
2 2013-11-01        3          3
3 2013-11-01        4          3
4 2013-11-01        5          3
timestamp    2013-11-01 00:00:00
grid_id                        1
row_count                      3
Name: 0, dtype: object


In [28]:
df[
    (df["timestamp"] == example["timestamp"]) &
    (df["grid_id"] == example["grid_id"])
]

,timestamp,grid_id,country_code,sms_in,sms_out,call_in,call_out,internet_activity,date,hour,day_of_week
0,2013-11-01,1,0,0.3521,NaN,NaN,0.0273,NaN,2013-11-01,0,Friday
1,2013-11-01,1,33,NaN,NaN,NaN,NaN,0.0261,2013-11-01,0,Friday
2,2013-11-01,1,39,1.7322,1.1047,0.5919,0.4020,57.7729,2013-11-01,0,Friday


In [29]:
df["total_sms"] = df["sms_in"] + df["sms_out"]

df["total_calls"] = df["call_in"] + df["call_out"]

df["total_activity"] = (
    df["total_sms"]
    + df["total_calls"]
    + df["internet_activity"]
)

In [30]:
print(
    df[
        [
            "total_sms",
            "total_calls",
            "total_activity"
        ]
    ].head()
)

   total_sms  total_calls  total_activity
0        NaN          NaN             NaN
1        NaN          NaN             NaN
2     2.8369       0.9939         61.6037
3        NaN          NaN             NaN
4        NaN          NaN             NaN


In [31]:
hourly_activity = (
    df.groupby("timestamp")[
        activity_cols
    ]
    .sum()
)

In [32]:
hourly_activity["total_activity"] = hourly_activity.sum(axis=1)

In [33]:
busiest_hour = hourly_activity["total_activity"].idxmax()

print(busiest_hour)
print(hourly_activity.loc[busiest_hour])

2013-11-01 11:00:00
sms_in               3.375077e+05
sms_out              1.743619e+05
call_in              2.372185e+05
call_out             2.847081e+05
internet_activity    4.151907e+06
total_activity       5.185704e+06
Name: 2013-11-01 11:00:00, dtype: float64


In [34]:
grid_activity = (
    df.groupby("grid_id")[
        activity_cols
    ]
    .sum()
)

grid_activity["total_activity"] = grid_activity.sum(axis=1)

In [35]:
busiest_grid = grid_activity["total_activity"].idxmax()

print(busiest_grid)
print(grid_activity.loc[busiest_grid])

5161
sms_in                16448.3863
sms_out                5663.4699
call_in                7650.4912
call_out               7807.5600
internet_activity    237230.3882
total_activity       274800.2956
Name: 5161, dtype: float64


In [36]:
null_counts = df.isna().sum()

print(null_counts)

timestamp                  0
grid_id                    0
country_code               0
sms_in               1086153
sms_out              1422446
call_in              1407781
call_out             1037413
internet_activity    1087074
date                       0
hour                       0
day_of_week                0
total_sms            1502892
total_calls          1594716
total_activity       1658462
dtype: int64


In [37]:
null_report = pd.DataFrame({
    "column": df.columns,
    "null_count": df.isna().sum().values
})

print(null_report)

               column  null_count
0           timestamp           0
1             grid_id           0
2        country_code           0
3              sms_in     1086153
4             sms_out     1422446
5             call_in     1407781
6            call_out     1037413
7   internet_activity     1087074
8                date           0
9                hour           0
10        day_of_week           0
11          total_sms     1502892
12        total_calls     1594716
13     total_activity     1658462


In [38]:
num_rows = len(df)
num_grids = df["grid_id"].nunique()
num_timestamps = df["timestamp"].nunique()
min_timestamp = df["timestamp"].min()
max_timestamp = df["timestamp"].max()
num_country_codes = df["country_code"].nunique()

print("Rows:", num_rows)
print("Unique grids:", num_grids)
print("Distinct timestamps:", num_timestamps)
print("Time range:", min_timestamp, "to", max_timestamp)
print("Country-code categories:", num_country_codes)

Rows: 1891928
Unique grids: 10000
Distinct timestamps: 24
Time range: 2013-11-01 00:00:00 to 2013-11-01 23:00:00
Country-code categories: 246


### Network Analytics Team — Data Profiling Summary

* The supplied file contains **1,891,928 raw activity records** covering **10,000 geographic grid cells** and **24 hourly timestamps** from **2013-11-01 00:00 to 23:00**. The timestamps have a verified **one-hour cadence**.

* The raw data grain is **timestamp + grid_id + country_code** rather than simply timestamp + grid_id. For example, grid 1 at 00:00 appears in three records with country codes **0, 33, and 39**. The country-code dimension must therefore be aggregated before producing the hourly grid analytics layer.

* The activity fields contain substantial blank values: `sms_in` 1,086,153; `sms_out` 1,422,446; `call_in` 1,407,781; `call_out` 1,037,413; and `internet_activity` 1,087,074. These blanks must be handled through the documented curated-layer null policy rather than silently removing otherwise valid records.

* Data-quality checks found **no missing grid IDs or timestamps, no exact duplicate records, and no negative activity values**. All observed grid IDs fall within the expected **1–10,000** range. The dataset contains **246 country-code categories**.

* After aggregating activity across country-code rows, the highest overall hourly activity occurred at **11:00**, with approximately **5.19 million total activity units**, while **grid 5161** had the highest cumulative activity at approximately **274,800 activity units** for the supplied day. These values represent communication activity measures and should not be interpreted as literal message counts, call counts, megabytes, or network congestion.
